blk/components/virt.c

```diff
@@ -0,0 +1,345 @@
+/* Fixed size memory allocator */
+static fsmem_t fsmem_data;
+static bitarray_t fsmem_avail_bitarr;
+static word_t fsmem_avail_bitarr_words[roundup_bits2words64(DRV_MAX_DATA_BUFFERS)];
+
+/* Datastore */
+static datastore_t ds;
+typedef struct ds_data {
+    uint32_t cli_id;
+    uint32_t cli_req_id;
+    uintptr_t cli_addr;
+    uintptr_t drv_addr;
+    uint16_t count;
+    blk_request_code_t code;
+} ds_data_t;
+static ds_data_t ds_data[DATASTORE_SIZE];
+static uint64_t ds_nextfree[DATASTORE_SIZE];
+static bool ds_used[DATASTORE_SIZE];
+
+/* Master boot record */
+struct mbr mbr;
+
+bool initialised = false;
+
+static void partitions_init() {
+    if (mbr.signature != MBR_SIGNATURE) {
+        LOG_BLK_VIRT_ERR("Invalid MBR signature\n");
+        return;
+    }
+
+    //@ericc: Figure out a better way to assign partitions to clients
+    int client_idx = 0;
+    int num_parts = 0;
+    for (int i = 0; i < MBR_MAX_PRIMARY_PARTITIONS; i++) {
+        if (mbr.partitions[i].type == MBR_PARTITION_TYPE_EMPTY) {
+            continue;
+        } else {
+            num_parts++;
+        }
+        
+        if (client_idx < BLK_NUM_CLIENTS) {
+            clients[client_idx].start_sector = mbr.partitions[i].lba_start;
+            clients[client_idx].sectors = mbr.partitions[i].sectors;
+            client_idx++;
+        }
+
+        if (mbr.partitions[i].lba_start % (BLK_TRANSFER_SIZE / MBR_SECTOR_SIZE) != 0) {
+            LOG_BLK_VIRT_ERR("Partition %d start sector %d not aligned to sDDF transfer size\n", i, mbr.partitions[i].lba_start);
+            return;
+        }
+    }
+
+    if (num_parts < BLK_NUM_CLIENTS) {
+        LOG_BLK_VIRT_ERR("Not enough partitions to assign to clients\n");
+        return;
+    }
+    
+    ((blk_storage_info_t *)blk_config)->sector_size = ((blk_storage_info_t *)blk_config_driver)->sector_size;
+    ((blk_storage_info_t *)blk_config)->block_size = ((blk_storage_info_t *)blk_config_driver)->block_size;
+    ((blk_storage_info_t *)blk_config)->size = clients[0].sectors / (BLK_TRANSFER_SIZE / MBR_SECTOR_SIZE);
+    ((blk_storage_info_t *)blk_config)->read_only = false;
+    ((blk_storage_info_t *)blk_config)->ready = true;
+#if BLK_NUM_CLIENTS > 1
+    ((blk_storage_info_t *)blk_config2)->sector_size = ((blk_storage_info_t *)blk_config_driver)->sector_size;
+    ((blk_storage_info_t *)blk_config2)->block_size = ((blk_storage_info_t *)blk_config_driver)->block_size;
+    ((blk_storage_info_t *)blk_config2)->size = clients[1].sectors / (BLK_TRANSFER_SIZE / MBR_SECTOR_SIZE);
+    ((blk_storage_info_t *)blk_config2)->read_only = false;
+    ((blk_storage_info_t *)blk_config2)->ready = true;
+#endif
+}
+
+static void request_mbr() {
+    uintptr_t mbr_addr;
+    fsmem_alloc(&fsmem_data, &mbr_addr, 1);
+    
+    uint64_t mbr_req_id;
+    ds_data_t mbr_req_data = {0, 0, 0, mbr_addr, 1, 0};
+    datastore_alloc(&ds, &mbr_req_data, &mbr_req_id);
+    
+    blk_enqueue_req(&drv_h, READ_BLOCKS, mbr_addr, 0, 1, mbr_req_id);
```

Comment:
We should always assert that an enqueue/dequeue is successful. In this case, we don't need an error check since this function should only be called in initialisation, but we should comment that.